tianshou/policy/modelfree/pg.py

```diff
@@ -190,18 +190,19 @@ def forward(
         """
         # TODO: rename? It's not really logits and there are particular
         #  assumptions about the order of the output and on distribution type
-        logits, hidden = self.actor(batch.obs, state=state, info=batch.info)
-        if isinstance(logits, tuple):
-            dist = self.dist_fn(*logits)
+        # See https://pytorch.org/docs/stable/distributions.html
+        action_dist_input, hidden = self.actor(batch.obs, state=state, info=batch.info)
+        if isinstance(action_dist_input, tuple):
+            dist = self.dist_fn(*action_dist_input)
         else:
-            dist = self.dist_fn(logits)
```

Comment:
Also i am surprised to call it `logits`, in the Pytorch documentation https://pytorch.org/docs/stable/distributions.html#categorical the first positional argument is always `probs`.

Reply:
Note, however, that the PyTorch class can also take `logits` as an argument (for the case where the inputs are unnormalised log probabilities) - and some implementations do pass `logits` explicitly (using a keyword argument).

Notably, the documentation in PGPolicy has led me to implement the default categorical distribution factory in the high-level API using `logits` instead of `probs`. If this is, in fact, not appropriate, then we should change the default. In particular, if there are activation functions at play that prevent the required values from being output, then the use of one instead of the other can have a fatal effect.

Reply:
But here

https://github.com/thu-ml/tianshou/blob/69a0e6bbd788509fe26bd93e142d0e78228114d6/test/discrete/test_pg.py#L72-L77

and here

https://github.com/thu-ml/tianshou/blob/69a0e6bbd788509fe26bd93e142d0e78228114d6/test/continuous/test_npg.py#L97-L105

are the places where i found test distributions ([Normal](https://pytorch.org/docs/stable/distributions.html#normal) and [Categorical](https://pytorch.org/docs/stable/distributions.html#categorical)), in both cases the first positional argument will be used here:

https://github.com/thu-ml/tianshou/blob/69a0e6bbd788509fe26bd93e142d0e78228114d6/tianshou/policy/modelfree/pg.py#L193-L197

The high level implementation is this one right?

https://github.com/thu-ml/tianshou/blob/69a0e6bbd788509fe26bd93e142d0e78228114d6/tianshou/highlevel/agent.py#L244-L278

Where can I see you had to use `logits` instead of `probs` ? i assume the `dist_fn` is passed in the `**kwargs` of line 277.

Sorry if there is something confusing, i am trying to understand better your last comment. Thanks :)

Reply:
In the high-level API, the starting point is the parameter object for PG, which contains a `dist_fn` parameter.

https://github.com/thu-ml/tianshou/blob/4756ee80ff11cd8692aef3752f35c0af60a452e8/tianshou/highlevel/params/policy_params.py#L310-L331

By default, the parameter will be an instance of `DistributionFunctionFactoryDefault` (as described in the docstring),

https://github.com/thu-ml/tianshou/blob/4756ee80ff11cd8692aef3752f35c0af60a452e8/tianshou/highlevel/params/dist_fn.py#L36-L44

and for the discrete case, you get `DistributionFunctionFactoryCategorical`, which is implemented like this:

https://github.com/thu-ml/tianshou/blob/4756ee80ff11cd8692aef3752f35c0af60a452e8/tianshou/highlevel/params/dist_fn.py#L16-L23

Note that I didn't "have to" use logits. I used logits in this implementation only because that's what the docstring in `PGPolicy` suggested to me:

https://github.com/thu-ml/tianshou/blob/4756ee80ff11cd8692aef3752f35c0af60a452e8/tianshou/policy/modelfree/pg.py#L46-L51

Reply:
Resolving this, part of the naming can't be improved now because we don't have Algorithm yet